onboarding/mlsys/1-gradients.md

```diff
@@ -0,0 +1,494 @@
+in algebra they would there's no constant term.  Just a straight line through
+the origin.
+
+## The Data Set
+
+Our made-up dataset has 4 data points (`true` indicates "ground truth"):
+
+- `(x=1, ytrue=1)`
+- `(x=2, ytrue=2)`
+- `(x=3, ytrue=3)`
+- `(x=4, ytrue=4)`
+
+## The Math Problem
+
+What is the best value of `m` to make our model (`ypred = m*x`) most accurately
+predict our dataset?
+
+### Q1:
+
+The answer is `m=1`.  Looking at the dataset, explain why `m=1` is correct
+before proceeding with further math.
+
+## What does "most accurately" mean, really?
+
+For this model and dataset, there is a perfect answer (`m=1`), but usually there
+isn't.  So usually instead of seeking an perfect answer, we are really seeking
+to minimize a function, called the "loss", that mathematically describes the
+goodness-of-fit between our model and the data.  There are lots of loss
+functions, but we're going to work with the Mean Squared Error (MSE) function.
+
+The MSE definition is just like it sounds: the mean of the squares of the
+errors between each ground truth `ytrue` and the corresponding model prediction
+`ypred`.  So for `N` data points, we can say:
+
+    loss_mse = ( (ytrue0 - ypred0)**2 + ... + (ytrueN - ypredN)**2 ) / N
+
+It should be clear from the above formula that the number of data points in the
+MSE loss calculation `N` is just a constant divisor that applies to each term.
+Then the MSE becomes the sum of a losses calculated individually for each
+point:
+
+    loss_point = (ytrue - ypred)**2 / N
+
+And we can rewrite `loss_point` in terms of `m`:
+
+    loss_point = (ytrue - m * x)**2 / N
+
+Did you notice that the loss is a function of `m`, even though our model
+defines `ypred` as a function of `x`?  Also, our dataset
+`x` and `ypred` values appear in the loss calculation as constants, even though
+we might normally think of `x` as our variable.  That is because every `m`
+value on our loss curve corresponds to a different model for `ypred = m*x`.
+The value of the loss at each value `m` is defined by how accurately it
+predicts the (constant) values in our dataset.
+
+![multi](multi.jpg "Multi Graph")
+
+### Q2:
+
+Can `loss_point` or `loss_mse` ever be negative?  Why or why not, and does the
+make logical sense to you?
+
+### Q3:
+
+With `m=2`, complete the following table by calculating the loss for each
+point of our dataset:
+
+|  `x`  | `ytrue` | `ypred` | `loss_point` |
+| ----- | ------- | ------- | ------------ |
+|   1   |    1    |         |              |
+|   2   |    2    |         |              |
+|   3   |    3    |         |              |
+|   4   |    4    |         |              |
+
+Also calculate the total `loss_mse` for the whole dataset:
+
+| `loss_mse` |
+| ---------- |
+|            |
+
+### Q4:
+
+Write a function for calculating the total `loss_mse`, which takes as input
+the current model weight `m` and a list `(x, ytrue)` data points, and returns
```

Comment:
missing a word

...and a list (x, ytrue) of data points...

Reply:
reworded